stl/inc/ranges

```diff
@@ -2635,6 +2589,415 @@ namespace ranges {
 
+    // CLASS TEMPLATE ranges::join_view
+    // clang-format off
+    template <input_range _Vw>
+        requires view<_Vw> && input_range<range_reference_t<_Vw>>
+            && (is_reference_v<range_reference_t<_Vw>> || view<range_value_t<_Vw>>)
+    class join_view;
+    // clang-format on
+
+    template <class _Vw>
+    class _Join_view_base : public view_interface<join_view<_Vw>> {
+    protected:
+        /* [[no_unique_address]] */ views::all_t<range_reference_t<_Vw>> _Inner{};
+    };
+
+    // clang-format off
+    template <class _Vw>
+        requires is_reference_v<range_reference_t<_Vw>>
+    class _Join_view_base<_Vw> : public view_interface<join_view<_Vw>> {};
+
+    template <input_range _Vw>
+        requires view<_Vw>&& input_range<range_reference_t<_Vw>>
+            && (is_reference_v<range_reference_t<_Vw>> || view<range_value_t<_Vw>>)
+    class join_view : public _Join_view_base<_Vw> {
+        // clang-format on
+    private:
+        using _InnerRng = range_reference_t<_Vw>;
+        /* [[no_unique_address]] */ _Vw _Range{};
+
+        template <bool _Const>
+        class _Sentinel;
+
+        template <class _OuterTraits, class _InnerTraits, bool _Deref_is_glvalue> // TRANSITION, LWG-3289
+        struct _Category_base {};
+
+        // clang-format off
+        template <class _OuterTraits, class _InnerTraits, bool _Deref_is_glvalue>
+            requires _Has_member_iterator_category<_InnerTraits> && _Has_member_iterator_category<_OuterTraits>
+        struct _Category_base<_OuterTraits, _InnerTraits, _Deref_is_glvalue> {
+            using iterator_category =
+                conditional_t<_Deref_is_glvalue
+                           && derived_from<typename _OuterTraits::iterator_category, bidirectional_iterator_tag>
+                           && derived_from<typename _InnerTraits::iterator_category, bidirectional_iterator_tag>,
+                    bidirectional_iterator_tag,
+                    conditional_t<_Deref_is_glvalue
+                               && derived_from<typename _OuterTraits::iterator_category, forward_iterator_tag>
+                               && derived_from<typename _InnerTraits::iterator_category, forward_iterator_tag>,
+                        forward_iterator_tag,
+                        conditional_t<derived_from<typename _OuterTraits::iterator_category, input_iterator_tag>
+                                   && derived_from<typename _InnerTraits::iterator_category, input_iterator_tag>,
+                            input_iterator_tag,
+                            output_iterator_tag>>>;
+        };
+        // clang-format on
+
+        template <bool _Const>
+        class _Iterator : public _Category_base<iterator_traits<iterator_t<_Maybe_const<_Const, _Vw>>>,
+                              iterator_traits<iterator_t<range_reference_t<_Maybe_const<_Const, _Vw>>>>,
+                              is_reference_v<range_reference_t<_Maybe_const<_Const, _Vw>>>> {

```

Comment:
Per WG21-P2259, the `iterator_category` is defined if and only if `_Deref_is_glvalue` is true, `_Base` models `forward_range`, and `range_reference_t<_Base>` models `forward_range`.

Maybe something like:
```c++
template <class _OuterRange,
          class _InnerRange = iterator_t<range_reference_t<_OuterRange>>,
          bool _Deref_is_glvalue =
              is_reference_v<range_reference_t<_OuterRange>>>
struct _Category_base {};

template <forward_range _OuterRange, forward_range _InnerRange>
struct _Category_base<_OuterRange, _InnerRange, true> {
  using _OuterCat = iterator_traits<iterator_t<_OuterRange>>::iterator_category;
  using _InnerCat = iterator_traits<iterator_t<_InnerRange>>::iterator_category;
  using iterator_category = conditional_t<
      derived_from<_OuterCat, bidirectional_iterator_tag> &&
          derived_from<_InnerCat, bidirectional_iterator_tag>,
      bidirectional_iterator_tag,
      conditional_t<derived_from<_OuterCat, forward_iterator_tag> &&
                        derived_from<_InnerCat, forward_iterator_tag>,
                    forward_iterator_tag, input_iterator_tag>>;
};

template <bool _Const>
class _Iterator : public _Category_base<iterator_t<_Maybe_const<_Const, _Vw>>> {
```

Reply:
@CaseyCarter, Should we already adopt this here or should this be a separat PR because the paper is a bit on larger side?

Reply:
I plan to implement WG21-P2259 separately after this and `split_view` land.